### PR TITLE
Nits in `IntegerOverflowError` and `RestrictedExprError`

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

* Added doc strings & made minimal changes to the error messages in `IntegerOverflowError`.
* Added a new field to `RestrictedExprError::InvalidRestrictedExpression` that records the (sub-)expression that is the source of the error. I _believe_ this is a non-breaking change because `RestrictedExprError` is never directly `pub use`d in `cedar-policy`, so the structure of the the variants shouldn't be visible to external users. Also made minor tweaks to comments and error messages.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [X] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by shaobo-he-aws — APPROVED** [review]
LGTM.

File: cedar-policy-core/src/ast/restricted_expr.rs [hunk 1]
```diff
@@ -309,16 +320,20 @@ impl<'a> Hash for RestrictedExprShapeOnly<'a> {
...
+    /// a feature that is not allowed in restricted expressions. The `feature`
+    /// argument is a string description of the feature that is not allowed.
+    /// The `expr` argument is the expression that uses the disallowed feature.
+    /// Note that it is potentially a sub-expression of a larger expression.
+    #[error("not allowed to use {feature} in a restricted expression: {expr}")]
     InvalidRestrictedExpression {
         /// what disallowed feature appeared in the expression
         feature: String,
+        /// the (sub-)expression that uses the disallowed feature
+        expr: Expr,
```
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/restricted_expr.rs:336]: while we're changing the fields of this variant anyway, want to change `String` to `SmolStr`?

**Review by cdisselkoen — APPROVED** [review]
LGTM

File: cedar-policy-core/src/ast/restricted_expr.rs (hunk 1 above)
Comment (khieta) [reply cedar-policy-core/src/ast/restricted_expr.rs]: Sure. Done in the latest commit.

**Comment by khieta** [comment]
Downstream build failure is due to #235, not this PR

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**